dbms/src/Flash/Coprocessor/DAGExpressionAnalyzer.cpp

```diff
@@ -192,6 +200,150 @@ void appendWindowDescription(
     window_columns.emplace_back(func_string, result_type);
     source_columns.emplace_back(func_string, result_type);
 }
+
+void setAuxiliaryColumnInfoImpl(
+    WindowDescription & window_desc,
+    const String & aux_col_name,
+    const Block & tmp_block,
+    bool is_begin)
+{
+    if (!aux_col_name.empty())
+    {
+        // Prepare something
+        Int32 & range_auxiliary_column_index = is_begin ? window_desc.frame.begin_range_auxiliary_column_index : window_desc.frame.end_range_auxiliary_column_index;
+        TypeIndex & aux_col_type = is_begin ? window_desc.begin_aux_col_type : window_desc.end_aux_col_type;
+
+        // Set auxiliary columns' indexes
+        size_t aux_col_idx = tmp_block.getPositionByName(aux_col_name);
+        range_auxiliary_column_index = aux_col_idx;
+
+        // Set auxiliary columns' types
+        const auto & col_and_name = tmp_block.getByName(aux_col_name);
+        auto data_type = col_and_name.type;
+        RUNTIME_CHECK_MSG(!data_type->isNullable(), "Get an unexpected nullable column when the frame type is range");
```

Comment:
> Why nullable type is not allowed?

Because MySQL forbids to input a nullable order by column.

Reply:
```
 explain select *, first_value(v) over (partition by p order by (case 0 is null when true then o else null end) range between 3 preceding and 0 following) as a from test.first_range;
+----------------------------------+---------+-----------+-------------------+---------------------------------------------------------------------------------------------------------------------------------------------+
| id                               | estRows | task      | access object     | operator info                                                                                                                               |
+----------------------------------+---------+-----------+-------------------+---------------------------------------------------------------------------------------------------------------------------------------------+
| Projection_6                     | 17.00   | root      |                   | test.first_range.p, test.first_range.o, test.first_range.v, Column#7                                                                        |
| └─Shuffle_13                     | 17.00   | root      |                   | execution info: concurrency:5, data sources:[Projection_9]                                                                                  |
|   └─Window_7                     | 17.00   | root      |                   | first_value(test.first_range.v)->Column#7 over(partition by test.first_range.p order by Column#6 range between 3 preceding and 0 following) |
|     └─Sort_12                    | 17.00   | root      |                   | test.first_range.p, Column#6                                                                                                                |
|       └─Projection_9             | 17.00   | root      |                   | test.first_range.p, test.first_range.o, test.first_range.v, <nil>->Column#6                                                                 |
|         └─TableReader_11         | 17.00   | root      |                   | data:TableFullScan_10                                                                                                                       |
|           └─TableFullScan_10     | 17.00   | cop[tikv] | table:first_range | keep order:false, stats:pseudo                                                                                                              |
+----------------------------------+---------+-----------+-------------------+---------------------------------------------------------------------------------------------------------------------------------------------+
```
It looks like `Column#6` is an only null column const

Reply:
> ```
>  explain select *, first_value(v) over (partition by p order by (case 0 is null when true then o else null end) range between 3 preceding and 0 following) as a from test.first_range;
> +----------------------------------+---------+-----------+-------------------+---------------------------------------------------------------------------------------------------------------------------------------------+
> | id                               | estRows | task      | access object     | operator info                                                                                                                               |
> +----------------------------------+---------+-----------+-------------------+---------------------------------------------------------------------------------------------------------------------------------------------+
> | Projection_6                     | 17.00   | root      |                   | test.first_range.p, test.first_range.o, test.first_range.v, Column#7                                                                        |
> | └─Shuffle_13                     | 17.00   | root      |                   | execution info: concurrency:5, data sources:[Projection_9]                                                                                  |
> |   └─Window_7                     | 17.00   | root      |                   | first_value(test.first_range.v)->Column#7 over(partition by test.first_range.p order by Column#6 range between 3 preceding and 0 following) |
> |     └─Sort_12                    | 17.00   | root      |                   | test.first_range.p, Column#6                                                                                                                |
> |       └─Projection_9             | 17.00   | root      |                   | test.first_range.p, test.first_range.o, test.first_range.v, <nil>->Column#6                                                                 |
> |         └─TableReader_11         | 17.00   | root      |                   | data:TableFullScan_10                                                                                                                       |
> |           └─TableFullScan_10     | 17.00   | cop[tikv] | table:first_range | keep order:false, stats:pseudo                                                                                                              |
> +----------------------------------+---------+-----------+-------------------+---------------------------------------------------------------------------------------------------------------------------------------------+
> ```
> 
> It looks like `Column#6` is an only null column const

okk, we will raise error when the column is nullable.

Reply:
> okk, we will raise error when the column is nullable.
So is it expected that an error will be reported when executing this query?

Reply:
> ```
>  explain select *, first_value(v) over (partition by p order by (case 0 is null when true then o else null end) range between 3 preceding and 0 following) as a from test.first_range;
> +----------------------------------+---------+-----------+-------------------+---------------------------------------------------------------------------------------------------------------------------------------------+
> | id                               | estRows | task      | access object     | operator info                                                                                                                               |
> +----------------------------------+---------+-----------+-------------------+---------------------------------------------------------------------------------------------------------------------------------------------+
> | Projection_6                     | 17.00   | root      |                   | test.first_range.p, test.first_range.o, test.first_range.v, Column#7                                                                        |
> | └─Shuffle_13                     | 17.00   | root      |                   | execution info: concurrency:5, data sources:[Projection_9]                                                                                  |
> |   └─Window_7                     | 17.00   | root      |                   | first_value(test.first_range.v)->Column#7 over(partition by test.first_range.p order by Column#6 range between 3 preceding and 0 following) |
> |     └─Sort_12                    | 17.00   | root      |                   | test.first_range.p, Column#6                                                                                                                |
> |       └─Projection_9             | 17.00   | root      |                   | test.first_range.p, test.first_range.o, test.first_range.v, <nil>->Column#6                                                                 |
> |         └─TableReader_11         | 17.00   | root      |                   | data:TableFullScan_10                                                                                                                       |
> |           └─TableFullScan_10     | 17.00   | cop[tikv] | table:first_range | keep order:false, stats:pseudo                                                                                                              |
> +----------------------------------+---------+-----------+-------------------+---------------------------------------------------------------------------------------------------------------------------------------------+
> ```
> 
> It looks like `Column#6` is an only null column const

I think we should raise error when the column is nullable as it's meaningless to calculate range frame with a nullable column.

Reply:
Will tidb report an error when executing such SQL?

Reply:
```
mysql> select *, first_value(v) over (partition by p order by (case 0 is null when true then o else null end) range between 3 preceding and 0 following) as a from test.first_range;
+---+----+----+---+
| p | o  | v  | a |
+---+----+----+---+
| 0 |  0 |  0 | 0 |
| 1 |  1 |  1 | 1 |
| 1 |  2 |  2 | 1 |
| 1 |  4 |  4 | 1 |
| 1 |  8 |  8 | 1 |
| 3 |  1 |  1 | 1 |
| 3 |  3 |  3 | 1 |
| 3 |  5 |  5 | 1 |
| 3 |  9 |  9 | 1 |
| 3 | 15 | 15 | 1 |
| 3 | 20 | 20 | 1 |
| 3 | 31 | 31 | 1 |
| 2 |  0 |  0 | 0 |
| 2 |  3 |  3 | 0 |
| 2 | 10 | 10 | 0 |
| 2 | 13 | 13 | 0 |
| 2 | 15 | 15 | 0 |
+---+----+----+---+
```

Reply:
If you are sure not to consider the nullable column, please do not push such sql down to tiflash, and inform the user in warnings

Reply:
Looks like mysql does not throw error if the order by column is nullable or even contains null value, since both TiDB and MySQL does not throw error here, I think TiFlash should not throw error, we can either support it or as @SeaRise suggests, "do not push to tiflash and inform the user in warnings"